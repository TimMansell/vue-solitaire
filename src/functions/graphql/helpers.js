import { gql } from 'apollo-server-lambda';

// eslint-disable-next-line import/prefer-default-export
export const grabQuery = ({ body }) => {
  const { query } = JSON.parse(body);

  const gqlQuery = gql`
    ${query}
  `;

  return gqlQuery;
};

export const grabVariables = ({ body }) => {
  const { variables } = JSON.parse(body);

  return variables;
};
