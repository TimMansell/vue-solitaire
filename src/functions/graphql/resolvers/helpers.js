import { gql } from 'apollo-boost';

export const formatVariables = (args, params) => {
  const { id, data } = args;

  const variables = {
    id,
    data: {
      ...data,
      ...params,
    },
  };

  return variables;
};

export const updateGameMutation = async (client, variables) => {
  const mutation = gql`
    mutation UpdateAGame($id: ID!, $data: GameInput!) {
      updateGame(id: $id, data: $data) {
        _id
        gameNumber
      }
    }
  `;

  const body = await client.mutate({
    mutation,
    variables,
  });

  return body.data.updateGame;
};

export const newGameMutation = async (client, variables) => {
  const mutation = gql`
    mutation NewGame($date: String!) {
      newGame(date: $date) {
        _id
        gameNumber
      }
    }
  `;

  const body = await client.mutate({
    mutation,
    variables,
  });

  return body.data.newGame;
};
