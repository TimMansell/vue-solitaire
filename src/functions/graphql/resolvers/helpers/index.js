export {
  insertIntoDb,
  countItemsInDb,
  findAllItems,
  findItemsInDb,
  findItemInDb,
} from './db';
export { findLeaderboardItems } from './find';
export { createPlayerName } from './user';
