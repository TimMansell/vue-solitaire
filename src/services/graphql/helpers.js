export const formatResponse = (response) => ({
  error: false,
  response,
});

export const formatError = () => ({
  error: true,
});
