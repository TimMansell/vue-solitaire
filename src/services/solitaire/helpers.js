import shuffle from 'lodash.shuffle';

const shuffleCards = ({ values, suits }) => {
  const deck = values.flatMap((value, index) =>
    suits.map((suit) => {
      const card = {
        id: `${index}${suit}`,
        value,
        order: index + 1,
        suit,
        visible: false,
      };

      return card;
    })
  );

  const shuffledDeck = shuffle(deck);

  return shuffledDeck;
};

const showHideCards = (cards, offset = 0) =>
  cards.map((card, index) => {
    if ((index + offset) % 2 === 0) {
      return {
        ...card,
        visible: true,
      };
    }

    return card;
  });

const getSelectedCard = (cards, selectedCardId) => {
  const [selectedCard] = cards.flat().filter((card) => card.id === selectedCardId);

  return selectedCard;
};

const getSelectedCardPosition = (boardCards, selectedCardId) => {
  const columnNo = boardCards.findIndex((cards) =>
    cards.find((card) => card.id === selectedCardId)
  );

  const cardPosition = boardCards[columnNo].findIndex((card) => card.id === selectedCardId);

  return {
    columnNo,
    cardPosition,
  };
};

const getVisibleCards = (cards) => {
  const visibleCards = cards.flat().filter((card) => card.visible);

  return visibleCards;
};

const getLastCard = (board, selectedColumn) => {
  const [selectedCard] = board[selectedColumn].slice(-1);

  return selectedCard;
};

const getLastCards = (cards) => {
  const lastCards = cards.map((card) => card.slice(-1)).flat();

  return lastCards;
};

const moveCardsFrom = (selectedCardId, cards) => {
  const { columnNo, cardPosition } = getSelectedCardPosition(cards, selectedCardId);

  const columnCards = cards[columnNo].slice(0, cardPosition);

  const remainingCards = columnCards.map((card, index) => {
    if (index === columnCards.length - 1 && !card.visible) {
      const newValues = {
        ...card,
        visible: true,
      };
      return newValues;
    }

    return card;
  });

  return {
    column: columnNo,
    cards: remainingCards,
  };
};

const moveCardsTo = (selectedCardId, selectedColumn, cardsFrom, cardsTo) => {
  const { columnNo, cardPosition } = getSelectedCardPosition(cardsFrom, selectedCardId);

  const columnCards = cardsTo[selectedColumn];
  const moveCards = cardsFrom[columnNo].slice(cardPosition);

  const newColumn = [...columnCards, ...moveCards];

  return {
    column: selectedColumn,
    cards: newColumn,
  };
};

export {
  shuffleCards,
  showHideCards,
  getSelectedCard,
  getSelectedCardPosition,
  getLastCard,
  getLastCards,
  getVisibleCards,
  moveCardsFrom,
  moveCardsTo,
};
