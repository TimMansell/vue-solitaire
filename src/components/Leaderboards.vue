<template>
  <div data-test="leaderboard">
    <p v-if="name" data-test="leaderboard-name">
      Your player name is: {{ name }}
    </p>

    <div ref="scrollTo">
      <Filters>
        <Select
          v-model="filters.showBest"
          label="Best"
          :items="bestItems"
          @select="setBest"
          data-test="leaderboard-set-best"
        />

        <Select
          v-model="filters.limit"
          label="Top"
          :items="limitItems"
          @select="displayLimit"
          data-test="leaderboard-set-top"
        />
      </Filters>
    </div>

    <h2 data-test="leaderboards-heading">
      Top {{ limit }} {{ topHeading }} {{ best }}
    </h2>

    <ResponsiveTable
      :headings="tableHeadings"
      :items="formattedLeaderboards"
      :placeholder-rows="limit"
      :to-highlight="{ key: 'player', value: name }"
    />

    <small v-if="filters.showBest === 'winPercent'">
      * Minimum of 25 games played
    </small>
  </div>
</template>

<script>
import xss from 'xss';
import { mapGetters, mapActions } from 'vuex';
import Filters from '@/components/Filters.vue';
import Select from '@/components/Select.vue';
import ResponsiveTable from '@/components/ResponsiveTable.vue';
import { formatDate } from '@/helpers/dates';

export default {
  name: 'Leaderboards',
  components: {
    Filters,
    Select,
    ResponsiveTable,
  },
  data() {
    return {
      bestItems: [
        { text: 'Moves', value: 'moves' },
        { text: 'Times', value: 'time' },
        { text: 'Win %', value: 'winPercent' },
        { text: 'Wins', value: 'wins' },
      ],
      limitItems: [
        { text: '25', value: 25 },
        { text: '50', value: 50 },
        { text: '100', value: 100 },
        { text: '500', value: 500 },
      ],
      filters: {
        limit: parseInt(xss(this.$route.params.limit), 10),
        showBest: xss(this.$route.params.showBest),
      },
    };
  },
  watch: {
    filters: {
      handler() {
        this.updateUrl();
      },
      deep: true,
    },
    $route: 'displayGames',
  },
  computed: {
    ...mapGetters(['leaderboards', 'name']),
    topHeading() {
      const { showBest } = this;

      const heading = {
        moves: 'Lowest',
        time: 'Quickest',
        winPercent: 'Best',
        wins: 'Most',
      };

      return heading[showBest];
    },
    formattedLeaderboards() {
      const { showBest, leaderboards } = this;

      const formatLeaderboardsContainingDate = (leaderboardsWithDate) =>
        leaderboardsWithDate.map((leaderboard) => ({
          ...leaderboard,
          date: formatDate(leaderboard.date),
        }));

      const format = {
        moves: () => formatLeaderboardsContainingDate(leaderboards),
        time: () => formatLeaderboardsContainingDate(leaderboards),
        winPercent: () => leaderboards,
        wins: () => leaderboards,
      };

      const formattedLeaderboard = format[showBest]?.() ?? leaderboards;

      return formattedLeaderboard;
    },
    tableHeadings() {
      const { showBest } = this;

      const defaultHeadings = ['', 'Player'];

      const headings = {
        moves: [...defaultHeadings, 'Date', 'Moves'],
        time: [...defaultHeadings, 'Date', 'Times'],
        winPercent: [...defaultHeadings, 'Win %'],
        wins: [...defaultHeadings, 'Wins'],
      };

      return headings[showBest];
    },
    showBest() {
      return this.filters.showBest;
    },
    limit() {
      return this.filters.limit;
    },
    best() {
      const [showBest] = this.bestItems.filter(
        ({ value }) => value === this.filters.showBest
      );

      return showBest?.text;
    },
  },
  mounted() {
    this.checkInitialFilters();
    this.displayGames();
  },
  destroyed() {
    this.clearLeaderboards();
  },
  methods: {
    ...mapActions(['getLeaderboards', 'clearLeaderboards']),
    checkInitialFilters() {
      const { limitItems, bestItems, limit, showBest } = this;

      const validLimit = limitItems.map(({ value }) => value).includes(limit);
      const validBest = bestItems.map(({ value }) => value).includes(showBest);

      this.filters.limit = validLimit ? limit : limitItems[0].value;
      this.filters.showBest = validBest ? showBest : bestItems[0].value;
    },
    displayLimit(limit) {
      this.filters.limit = parseInt(limit, 10);
    },
    setBest(best) {
      this.filters.showBest = best;
    },
    displayGames() {
      const { filters } = this;

      this.getLeaderboards(filters);
    },
    updateUrl() {
      const { limit, showBest } = this;

      this.$router.replace(`/leaderboards/${showBest}/${limit}`);
    },
  },
};
</script>
