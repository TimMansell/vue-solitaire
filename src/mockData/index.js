import { initBoard } from '../services/solitaire/board';

export const mockUid = 'f5c6a829-f0da-4dfc-81a0-e6419f0163c7';

export const mockDeck = [
  {
    id: 3,
    value: '3',
    order: 3,
    suit: '♣',
  },
  {
    id: 36,
    value: '10',
    order: 10,
    suit: '♥',
  },
  {
    id: 19,
    value: '6',
    order: 6,
    suit: '♠',
  },
  {
    id: 48,
    value: '9',
    order: 9,
    suit: '♦',
  },
  {
    id: 47,
    value: '8',
    order: 8,
    suit: '♦',
  },
  {
    id: 4,
    value: '4',
    order: 4,
    suit: '♣',
  },
  {
    id: 15,
    value: '2',
    order: 2,
    suit: '♠',
  },
  {
    id: 52,
    value: 'K',
    order: 13,
    suit: '♦',
  },
  {
    id: 42,
    value: '3',
    order: 3,
    suit: '♦',
  },
  {
    id: 43,
    value: '4',
    order: 4,
    suit: '♦',
  },
  {
    id: 39,
    value: 'K',
    order: 13,
    suit: '♥',
  },
  {
    id: 38,
    value: 'Q',
    order: 12,
    suit: '♥',
  },
  {
    id: 41,
    value: '2',
    order: 2,
    suit: '♦',
  },
  {
    id: 37,
    value: 'J',
    order: 11,
    suit: '♥',
  },
  {
    id: 2,
    value: '2',
    order: 2,
    suit: '♣',
  },
  {
    id: 34,
    value: '8',
    order: 8,
    suit: '♥',
  },
  {
    id: 11,
    value: 'J',
    order: 11,
    suit: '♣',
  },
  {
    id: 50,
    value: 'J',
    order: 11,
    suit: '♦',
  },
  {
    id: 16,
    value: '3',
    order: 3,
    suit: '♠',
  },
  {
    id: 25,
    value: 'Q',
    order: 12,
    suit: '♠',
  },
  {
    id: 7,
    value: '7',
    order: 7,
    suit: '♣',
  },
  {
    id: 23,
    value: '10',
    order: 10,
    suit: '♠',
  },
  {
    id: 21,
    value: '8',
    order: 8,
    suit: '♠',
  },
  {
    id: 18,
    value: '5',
    order: 5,
    suit: '♠',
  },
  {
    id: 46,
    value: '7',
    order: 7,
    suit: '♦',
  },
  {
    id: 29,
    value: '3',
    order: 3,
    suit: '♥',
  },
  {
    id: 12,
    value: 'Q',
    order: 12,
    suit: '♣',
  },
  {
    id: 22,
    value: '9',
    order: 9,
    suit: '♠',
  },
  {
    id: 30,
    value: '4',
    order: 4,
    suit: '♥',
  },
  {
    id: 8,
    value: '8',
    order: 8,
    suit: '♣',
  },
  {
    id: 28,
    value: '2',
    order: 2,
    suit: '♥',
  },
  {
    id: 45,
    value: '6',
    order: 6,
    suit: '♦',
  },
  {
    id: 27,
    value: 'A',
    order: 1,
    suit: '♥',
  },
  {
    id: 33,
    value: '7',
    order: 7,
    suit: '♥',
  },
  {
    id: 14,
    value: 'A',
    order: 1,
    suit: '♠',
  },
  {
    id: 6,
    value: '6',
    order: 6,
    suit: '♣',
  },
  {
    id: 9,
    value: '9',
    order: 9,
    suit: '♣',
  },
  {
    id: 13,
    value: 'K',
    order: 13,
    suit: '♣',
  },
  {
    id: 32,
    value: '6',
    order: 6,
    suit: '♥',
  },
  {
    id: 51,
    value: 'Q',
    order: 12,
    suit: '♦',
  },
  {
    id: 44,
    value: '5',
    order: 5,
    suit: '♦',
  },
  {
    id: 31,
    value: '5',
    order: 5,
    suit: '♥',
  },
  {
    id: 26,
    value: 'K',
    order: 13,
    suit: '♠',
  },
  {
    id: 40,
    value: 'A',
    order: 1,
    suit: '♦',
  },
  {
    id: 10,
    value: '10',
    order: 10,
    suit: '♣',
  },
  {
    id: 1,
    value: 'A',
    order: 1,
    suit: '♣',
  },
  {
    id: 24,
    value: 'J',
    order: 11,
    suit: '♠',
  },
  {
    id: 5,
    value: '5',
    order: 5,
    suit: '♣',
  },
  {
    id: 35,
    value: '9',
    order: 9,
    suit: '♥',
  },
  {
    id: 49,
    value: '10',
    order: 10,
    suit: '♦',
  },
  {
    id: 17,
    value: '4',
    order: 4,
    suit: '♠',
  },
  {
    id: 20,
    value: '7',
    order: 7,
    suit: '♠',
  },
];

export const mockBoard = initBoard(mockDeck);

export const mockFoundation = [
  [
    { id: 1, suit: '♠', value: 'A', visible: false },
    { id: 2, suit: '♠', value: 2, visible: true },
    { id: 3, suit: '♠', value: 3, visible: false },
    { id: 4, suit: '♠', value: 4, visible: true },
  ],
  [],
  [],
  [],
];

export const mockPlayerName = 'Player Name';

export const mockHistoryApi = [
  {
    date: '2021-05-20T23:34:49.564Z',
    won: false,
    lost: false,
    moves: 1,
    time: 12,
  },
  {
    date: '2021-05-19T23:34:49.564Z',
    won: true,
    lost: false,
    moves: 2,
    time: 12,
  },
  {
    date: '2021-05-19T23:34:49.564Z',
    won: false,
    lost: true,
    moves: 2,
    time: 12,
  },
  {
    date: '2021-05-19T23:34:49.564Z',
    won: true,
    lost: false,
    moves: 2,
    time: 12,
  },
];

export const mockHistory = [
  {
    date: '2021-05-20T23:34:49.564Z',
    time: '2021-05-20T23:34:49.564Z',
    duration: '0:00:12',
    moves: 1,
    number: 4,
    outcome: 'Gave Up',
  },
  {
    date: '2021-05-19T23:34:49.564Z',
    time: '2021-05-19T23:34:49.564Z',
    duration: '0:00:12',
    moves: 2,
    number: 3,
    outcome: 'Won',
  },
  {
    date: '2021-05-19T23:34:49.564Z',
    time: '2021-05-19T23:34:49.564Z',
    duration: '0:00:12',
    moves: 2,
    number: 2,
    outcome: 'Lost',
  },
  {
    date: '2021-05-19T23:34:49.564Z',
    time: '2021-05-19T23:34:49.564Z',
    duration: '0:00:12',
    moves: 2,
    number: 1,
    outcome: 'Won',
  },
];

export const mockLeaderboardsMovesAPI = [
  {
    date: '2021-04-29T12:25:47.907Z',
    uid: '7dac9d78-353f-409b-8a7f-2192409c44a2',
    moves: 2,
  },
  {
    date: '2021-04-29T12:26:20.825Z',
    uid: '2cbf658a-3102-4e9d-b749-bac853efed0d',
    moves: 2,
  },
];

export const mockLeaderboardsMovesLegacy = [
  {
    rank: 1,
    date: '29-04-2021',
    player: 'Player 1',
    moves: 2,
  },
  {
    rank: 2,
    date: '29-04-2021',
    player: 'Player 2',
    moves: 2,
  },
];

export const mockLeaderboardsMoves = [
  {
    rank: 1,
    date: '2021-04-29T12:25:47.907Z',
    player: 'Player 1',
    moves: 2,
  },
  {
    rank: 2,
    date: '2021-04-29T12:26:20.825Z',
    player: 'Player 2',
    moves: 2,
  },
];

export const mockLeaderboardsTimesAPI = [
  {
    date: '2021-04-29T12:25:47.907Z',
    uid: '7dac9d78-353f-409b-8a7f-2192409c44a2',
    time: 20,
  },
  {
    date: '2021-04-29T12:26:20.825Z',
    uid: '2cbf658a-3102-4e9d-b749-bac853efed0d',
    time: 200,
  },
];

export const mockLeaderboardsTimesLegacy = [
  {
    rank: 1,
    date: '29-04-2021',
    player: 'Player 1',
    time: '0:00:20',
  },
  {
    rank: 2,
    date: '29-04-2021',
    player: 'Player 2',
    time: '0:03:20',
  },
];

export const mockLeaderboardsTimes = [
  {
    rank: 1,
    date: '2021-04-29T12:25:47.907Z',
    player: 'Player 1',
    duration: '0:00:20',
  },
  {
    rank: 2,
    date: '2021-04-29T12:26:20.825Z',
    player: 'Player 2',
    duration: '0:03:20',
  },
];

export const mockPlayers = [
  { uid: '7dac9d78-353f-409b-8a7f-2192409c44a2', name: 'Player 1' },
  { uid: '2cbf658a-3102-4e9d-b749-bac853efed0d', name: 'Player 2' },
];

export const mockStats = { won: 2, lost: 4, completed: 9, players: 4 };

export const mockVersionNumber = 'v1.0.0';
