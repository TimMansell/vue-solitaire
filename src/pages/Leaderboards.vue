<template>
  <GameOverlay
    ref="leaderboardOverlay"
    center-content
    show-close
    data-test="leaderboards"
  >
    <template #title> Leaderboards </template>
    <template #msg>
      <Leaderboards @scrollTo="scrollTo" />
    </template>
  </GameOverlay>
</template>

<script>
import VueScrollTo from 'vue-scrollto';
import GameOverlay from '@/components/GameOverlay.vue';
import Leaderboards from '@/components/Leaderboards.vue';

export default {
  name: 'LeaderboardsPage',
  components: {
    GameOverlay,
    Leaderboards,
  },
  methods: {
    scrollTo(elem) {
      const { leaderboardOverlay } = this.$refs;

      VueScrollTo.scrollTo(elem, {
        container: leaderboardOverlay.$el,
        offset: -10,
      });
    },
  },
};
</script>
