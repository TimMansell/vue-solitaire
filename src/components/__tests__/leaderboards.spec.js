import { describe, it, expect, vi } from 'vitest';
import tzMock from 'timezone-mock';
import { shallowMount } from '@vue/test-utils';
import Leaderboards from '@/components/Leaderboards.vue';
import { mockUid, mockLeaderboardsMoves, mockPlayerName } from '@/mockData';
import { setupStore, setupRoute } from '@@/tests/helpers';

tzMock.register('UTC');

const global = {
  mocks: {
    $store: setupStore({
      dispatch: vi.fn(),
      leaderboards: mockLeaderboardsMoves,
      luid: mockUid,
      name: mockPlayerName,
    }),
    $route: setupRoute({
      showBest: 'moves',
      limit: 25,
    }),
  },
};

describe('Leaderboards.vue', () => {
  it('renders the component without crashing', () => {
    const wrapper = shallowMount(Leaderboards, {
      global,
    });

    expect(wrapper.isVisible()).toBe(true);
  });

  it('should set default filters when route params are incorrect format', () => {
    const wrapper = shallowMount(Leaderboards, {
      global: {
        mocks: {
          ...global.mocks,
          $route: setupRoute({
            showBest: 5000,
            limit: 'abc',
          }),
        },
      },
    });

    expect(wrapper.vm.filters).toStrictEqual({
      showBest: 'moves',
      limit: 25,
    });
  });

  it('should set default filters when route params are correct format but are out of bounds', () => {
    const wrapper = shallowMount(Leaderboards, {
      global: {
        mocks: {
          ...global.mocks,
          $route: setupRoute({
            showBest: 'time',
            limit: 5000,
          }),
        },
      },
    });

    expect(wrapper.vm.filters).toStrictEqual({
      showBest: 'time',
      limit: 25,
    });
  });

  it('should show correct top message', () => {
    const wrapper = shallowMount(Leaderboards, {
      global,
    });

    expect(wrapper.find('[data-test="leaderboards-heading"]').text()).toContain(
      'Top 25 Lowest Moves'
    );
  });

  it('should show correct table headings', () => {
    const wrapper = shallowMount(Leaderboards, {
      global,
    });

    expect(wrapper.vm.tableHeadings).toStrictEqual([
      '',
      'Player',
      'Date',
      'Moves',
    ]);
  });

  it('should show correct table items', () => {
    const wrapper = shallowMount(Leaderboards, {
      global,
    });

    expect(wrapper.vm.formattedLeaderboards).toStrictEqual([
      {
        date: '29-04-2021',
        moves: 2,
        player: 'Player 1',
        rank: 1,
      },
      {
        date: '29-04-2021',
        moves: 2,
        player: 'Player 2',
        rank: 2,
      },
    ]);
  });
});
