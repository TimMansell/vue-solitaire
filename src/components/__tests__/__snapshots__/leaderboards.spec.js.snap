// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Leaderboards.vue matches snapshot 1`] = `
<div data-test="leaderboard">
  <p data-test="leaderboard-name">
    Your player name is: Player Name
  </p>
  <div>
    <filters-stub>
      <select-stub label="Best" value="moves" items="[object Object],[object Object],[object Object],[object Object]" data-test="leaderboard-set-best"></select-stub>
      <select-stub label="Top" value="25" items="[object Object],[object Object],[object Object],[object Object]" data-test="leaderboard-set-top"></select-stub>
    </filters-stub>
  </div>
  <h2 data-test="leaderboards-heading">
    Top 25 Lowest Moves
  </h2>
  <responsivetable-stub headings=",Player,Date,Moves" items="[object Object],[object Object]" placeholderrows="25" tohighlight="[object Object]"></responsivetable-stub>
  <!---->
</div>
`;
